File: typescript/packages/providers/src/failover.ts
import type { ModelChunk, ModelProvider, StreamRequest } from '@agent-framework/core';

export interface FailoverProviderOptions {
  providers: ModelProvider[];
  maxAttempts?: number;
}

export class FailoverProvider implements ModelProvider {
  readonly id = 'failover';

  constructor(private readonly options: FailoverProviderOptions) {
    if (options.providers.length === 0) {
      throw new Error('FailoverProvider requires at least one provider');
    }
  }

  async *stream(request: StreamRequest): AsyncGenerator<ModelChunk, void, unknown> {
    const attempts = this.options.maxAttempts ?? this.options.providers.length;
    let lastError: Error | undefined;

    for (let index = 0; index < Math.min(attempts, this.options.providers.length); index += 1) {
      const provider = this.options.providers[index]!;
      try {
        for await (const chunk of provider.stream(request)) {
          yield chunk;
        }
        return;
      } catch (error) {
        lastError = error instanceof Error ? error : new Error(String(error));
      }
    }

    throw lastError ?? new Error('All providers failed');
  }
}

File: python/src/agent_framework/providers/failover.py
from __future__ import annotations

from typing import AsyncIterator

from agent_framework.core.domain import ModelChunk
from agent_framework.core.contracts import Message, ToolDefinition


class FailoverProvider:
    id = "failover"

    def __init__(self, providers: list, max_attempts: int | None = None) -> None:
        if not providers:
            raise ValueError("FailoverProvider requires at least one provider")
        self._providers = providers
        self._max_attempts = max_attempts or len(providers)

    async def stream(self, messages: list[Message], tools: list[ToolDefinition]) -> AsyncIterator[ModelChunk]:
        last_error: Exception | None = None
        for provider in self._providers[: self._max_attempts]:
            try:
                async for chunk in provider.stream(messages, tools):
                    yield chunk
                return
            except Exception as exc:  # noqa: BLE001
                last_error = exc
        raise last_error or RuntimeError("All providers failed")
